synapse/storage/databases/main/events_bg_updates.py

```diff
@@ -1399,3 +1406,83 @@ def drop_invalid_event_edges_txn(txn: LoggingTransaction) -> bool:
             )
 
         return batch_size
+
+    async def _background_events_populate_state_key_rejections(
+        self, progress: JsonDict, batch_size: int
+    ) -> int:
+        """Back-populate `events.state_key` and `events.rejection_reason"""
+
+        min_stream_ordering_exclusive = progress["min_stream_ordering_exclusive"]
+        max_stream_ordering_inclusive = progress["max_stream_ordering_inclusive"]
+
+        def _populate_txn(txn: LoggingTransaction) -> bool:
+            """Returns True if we're done."""
+
+            # first we need to find an endpoint.
+            # we need to find the final row in the batch of batch_size, which means
+            # we need to skip over (batch_size-1) rows and get the next row.
```

Comment:
`events` table can only have one row per stream ordering, so this isn't quite accurate.

Reply:
I don't really understand why it's not accurate.

suppose we have these rows in the db:

```
stream_ordering | event_id
----------------+-----------
 1              | $event1
 3              | $event2
 5              | $event3
 7              | $event4
 9              | $event5
```

... and we want to get a batch of size 3. So we need to skip over 2 rows and get the stream_ordering from the third.

Reply:
Oh, hrmmm, I misread the rationale I think. I was assuming we were doing this query to handle having multiple rows per stream ordering (which is why we often do it in other places for other tables. But actually the reason we're doing this is because we can't use `LIMIT` with `UPDATE`?

Reply:
even if we could use `LIMIT` with `UPDATE`, we'd need to know where we got up to, to persist back to the `progress` dict.